llvm/lib/SYCLLowerIR/LowerESIMDVecArg.cpp

```diff
@@ -0,0 +1,432 @@
+//        % "class._ZTSN2cm3gen4simdIiLi16EEE.cm::gen::simd" addrspace(4) *
+//        addrspacecast(% "class._ZTSN2cm3gen4simdIiLi16EEE.cm::gen::simd" *
+//        bitcast(<16 x i32> * @0 to
+//        %"class._ZTSN2cm3gen4simdIiLi16EEE.cm::gen::simd" *) to %
+//        "class._ZTSN2cm3gen4simdIiLi16EEE.cm::gen::simd" addrspace(4) *),
+//        i64 0, i32 0))
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "ESIMDLowerVecArg"
+
+namespace llvm {
+
+// Forward declarations
+void initializeESIMDLowerVecArgLegacyPassPass(PassRegistry &);
+ModulePass *createESIMDLowerVecArgPass();
+
+// Pass converts simd* function parameters and globals to
+// llvm's first-class vector* type.
+class ESIMDLowerVecArgPass {
+public:
+  bool run(Module &M);
+
+private:
+  DenseMap<GlobalVariable *, GlobalVariable *> OldNewGlobal;
+
+  Function *rewriteFunc(Function &F);
+  Type *argIsSimdPtr(Value *arg);
+  void fixGlobals(Module &M);
+  bool hasGlobalConstExpr(Value *V, GlobalVariable *&Global);
+  void replaceConstExprWithGlobals(Module &M);
+  ConstantExpr *createNewConstantExpr(GlobalVariable *newGlobalVar,
+                                      Type *oldGlobalType, Value *old);
+  void removeOldGlobals();
+};
+
+} // namespace llvm
+
+namespace {
+class ESIMDLowerVecArgLegacyPass : public ModulePass {
+public:
+  static char ID;
+  ESIMDLowerVecArgLegacyPass() : ModulePass(ID) {
+    initializeESIMDLowerVecArgLegacyPassPass(*PassRegistry::getPassRegistry());
+  }
+
+  bool runOnModule(Module &M) override {
+    auto Modified = Impl.run(M);
+    return Modified;
+  }
+
+  bool doInitialization(Module &M) override { return false; }
+
+private:
+  ESIMDLowerVecArgPass Impl;
+};
+} // namespace
+
+char ESIMDLowerVecArgLegacyPass::ID = 0;
+INITIALIZE_PASS(ESIMDLowerVecArgLegacyPass, "ESIMDLowerVecArg",
+                "Translate simd ptr to native vector type", false, false)
+
+// Public interface to VecArgPass
+ModulePass *llvm::createESIMDLowerVecArgPass() {
+  return new ESIMDLowerVecArgLegacyPass();
+}
+
+// Return ptr to first-class vector type if Value is a simd*, else return
+// nullptr.
+Type *ESIMDLowerVecArgPass::argIsSimdPtr(Value *arg) {
+  auto ArgType = arg->getType();
+  if (ArgType->isPointerTy()) {
+    auto containedType = ArgType->getPointerElementType();
+    if (containedType->isStructTy()) {
+      if (containedType->getStructNumElements() == 1 &&
+          containedType->getStructElementType(0)->isVectorTy()) {
+        return PointerType::get(containedType->getStructElementType(0),
+                                ArgType->getPointerAddressSpace());
+      }
+    }
+  }
+  return nullptr;
+}
+
+// F may have multiple arguments of type simd*. This
+// function updates all parameters along with call
+// call sites of F.
+Function *ESIMDLowerVecArgPass::rewriteFunc(Function &F) {
```

Comment:
Did you consider llvm::cloneFunctionInto()?

Reply:
thanks for pointing me to this, i wasnt aware of this function. i looked in to this now and it seems this does a deep copy. since we dont want to store old function, i think we can skip using this API.

Reply:
There are a lot of details that need to be handled to do this correctly, so I think it would be more maintainable to use the existing clone functions.

Reply:
updated patch to use llvm's CloneFunctionInto API.